datafusion/expr-common/src/type_coercion/binary.rs

```diff
@@ -68,11 +70,57 @@ impl Signature {
     }
 }
 
-/// Returns a [`Signature`] for applying `op` to arguments of type `lhs` and `rhs`
-fn signature(lhs: &DataType, op: &Operator, rhs: &DataType) -> Result<Signature> {
-    use arrow::datatypes::DataType::*;
-    use Operator::*;
-    match op {
+pub struct BinaryTypeCoercer<'a> {
+    lhs: &'a DataType,
+    op: &'a Operator,
+    rhs: &'a DataType,
+
+    lhs_spans: Spans,
```

Comment:
What about simply having a single `spans: Span` that is computed on construction? Is there any reason to keep around the different set of spans?

Reply:
They are also merged for the overarching error though, that's true